src/com/google/javascript/jscomp/PeepholeReplaceKnownMethods.java

```diff
@@ -956,4 +959,168 @@ private Node tryFoldStringSplit(Node n, Node stringNode, Node arg1) {
     reportChangeToEnclosingScope(parent);
     return arrayOfStrings;
   }
+
+  private Node tryToFoldArrayConcat(Node n) {
+    checkArgument(n.isCall(), n);
+
+    if (!isASTNormalized() || !useTypes) {
+      return n;
+    }
+    ConcatFunctionCall concatFunctionCall = createConcatFunctionCallForNode(n);
+    if (concatFunctionCall == null) {
+      return n;
+    }
+    if (isSafeToRemoveArrayLiteralFromFrontOfConcat(concatFunctionCall)) {
+      return removeArrayLiteralFromFrontOfConcat(concatFunctionCall);
+    }
+    if (isSafeToFoldConcatChaining(concatFunctionCall)) {
+      return foldConcatChaining(concatFunctionCall);
+    }
+    return n;
+  }
+
+  /**
+   * Simply replace empty array literal from the front of concatenation by the first argument of
+   * concat function call `[].concat(arr,1)` -> `arr.concat(1)`.
+   */
+  private Node removeArrayLiteralFromFrontOfConcat(ConcatFunctionCall concatFunctionCall) {
+    checkArgument(isSafeToRemoveArrayLiteralFromFrontOfConcat(concatFunctionCall),
+        concatFunctionCall);
+
+    Node callNode = concatFunctionCall.getCallNode();
+    Node firstArg = concatFunctionCall.getFirstArgumentNode();
+    callNode.removeChild(firstArg);
+    Node curentTarget = callNode.getFirstChild();
+    Node emptyArrayLiteral = concatFunctionCall.getCalleeNode();
+    curentTarget.replaceChild(emptyArrayLiteral, firstArg);
+    reportChangeToEnclosingScope(callNode);
+    return callNode;
+  }
+
+  /**
+   * Perform folding of chained concat functions, so `arr.concat(a).concat(b)` will be fold into
+   * `arr.concat(a,b)`;
+   */
+  private Node foldConcatChaining(ConcatFunctionCall concatFunctionCall) {
+    checkArgument(isSafeToFoldConcatChaining(concatFunctionCall), concatFunctionCall);
+
+    Node concatCall = concatFunctionCall.getCallNode();
+    Node previousConcatCall = concatFunctionCall.getCalleeNode();
+    Node arg = concatFunctionCall.getFirstArgumentNode();
+    while (arg != null) {
+      Node curentArg = arg;
+      arg = arg.getNext();
+      previousConcatCall.addChildToBack(curentArg.detach());
+    }
+    concatCall.replaceWith(previousConcatCall.detach());
+    reportChangeToEnclosingScope(previousConcatCall);
+    return previousConcatCall;
+  }
+
+  /**
+   * Check if we have this code pattern `[].concat(exactlyArrayArgument,...*)` so
+   * `removeArrayLiteralFromFrontOfConcat` can be performed.
+   */
+  private boolean isSafeToRemoveArrayLiteralFromFrontOfConcat(
```

Comment:
I would really prefer to have this logic in the same place as the logic doing the removal.

Maybe have a `tryToRemoveArrayLiteralFromFrontOfConcat(concatFunctionCall)` method?
It would return either the same concatFunctionCall it was given, or a new one for the newly constructed call.

Just thought of something though.
Do you intend to change `[].concat(myArray)` into just `myArray`?
I don't think I see logic to handle that case or a test case for it, but I don't think I see logic checking to make sure you don't end up getting `myArray.concat()` either.

Reply:
Done
No, this optimization not intended to change `[].concat(myArray)` into just` myArray`.
There is no need for special checks for this case. 
Because if `myarray` is array node or function which return array and only array, its possible to do this `[].concat(myArray)`->`myArray.concat()` those js code are equivalent (and that's all that we can optimize). 
(concat only takes one optional parameter https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat)
And 2 test cases exist for this purpose
https://github.com/qqingvarqq/closure-compiler/blob/c9520c0851b4f80b957250fa0743077403cddfc1/test/com/google/javascript/jscomp/PeepholeReplaceKnownMethodsTest.java#L625
https://github.com/qqingvarqq/closure-compiler/blob/c9520c0851b4f80b957250fa0743077403cddfc1/test/com/google/javascript/jscomp/PeepholeReplaceKnownMethodsTest.java#L627